core/src/main/java/org/apache/iceberg/deletes/RoaringPositionBitmap.java

```diff
@@ -0,0 +1,309 @@
+ *
+ * <p>Note that positions must be greater than or equal to 0 add less than {@link #MAX_POSITION}.
+ * This bitmap does not support positions larger than {@link #MAX_POSITION} as the bitmap array
+ * length is a signed 32-bit integer that must be greater than or equal to 0.
+ */
+class RoaringPositionBitmap {
+
+  static final long MAX_POSITION = toPosition(Integer.MAX_VALUE - 1, Integer.MIN_VALUE);
+  private static final RoaringBitmap[] EMPTY_BITMAP_ARRAY = new RoaringBitmap[] {};
+  private static final long BITMAP_COUNT_SIZE_BYTES = 8L;
+  private static final long BITMAP_KEY_SIZE_BYTES = 4L;
+
+  private RoaringBitmap[] bitmaps;
+
+  RoaringPositionBitmap() {
+    this.bitmaps = EMPTY_BITMAP_ARRAY;
+  }
+
+  private RoaringPositionBitmap(RoaringBitmap[] bitmaps) {
+    this.bitmaps = bitmaps;
+  }
+
+  /**
+   * Adds a position to the bitmap.
+   *
+   * @param pos the row position
+   */
+  public void add(long pos) {
+    validatePosition(pos);
+    int high = highBytes(pos);
+    int low = lowBytes(pos);
+    allocateBitmapsIfNeeded(high + 1 /* required number of bitmaps */);
+    bitmaps[high].add(low);
+  }
+
+  /**
+   * Adds a range of positions to the bitmap.
+   *
+   * @param posStartInclusive the start position of the range (inclusive)
+   * @param posEndExclusive the end position of the range (exclusive)
+   */
+  public void addRange(long posStartInclusive, long posEndExclusive) {
+    for (long pos = posStartInclusive; pos < posEndExclusive; pos++) {
+      add(pos);
+    }
+  }
+
+  /**
+   * Checks if the bitmap contains a position.
+   *
+   * @param pos the row position
+   * @return true if the position is in this bitmap, false otherwise
+   */
+  public boolean contains(long pos) {
+    validatePosition(pos);
+    int high = highBytes(pos);
+    int low = lowBytes(pos);
+    return high < bitmaps.length && bitmaps[high].contains(low);
+  }
+
+  /**
+   * Adds all positions from the other bitmap to this bitmap, modifying this bitmap in place.
+   *
+   * @param that the other bitmap
+   */
+  public void or(RoaringPositionBitmap that) {
+    allocateBitmapsIfNeeded(that.bitmaps.length);
+    for (int index = 0; index < that.bitmaps.length; index++) {
+      bitmaps[index].or(that.bitmaps[index]);
+    }
+  }
+
+  /**
+   * Indicates whether the bitmap is empty.
+   *
+   * @return true if the bitmap is empty, false otherwise
+   */
+  public boolean isEmpty() {
+    return cardinality() == 0;
+  }
+
+  /**
+   * Returns the number of positions in the bitmap.
+   *
+   * @return the number of set positions
+   */
+  public long cardinality() {
+    long cardinality = 0L;
+    for (RoaringBitmap bitmap : bitmaps) {
+      cardinality += bitmap.getLongCardinality();
+    }
+    return cardinality;
+  }
+
+  /**
+   * Run-length compresses the bitmap if it is more space efficient.
+   *
+   * @return whether the bitmap was compressed
+   */
+  public boolean runOptimize() {
```

Comment:
I'm not sure if we want to expose Roaring bitmap apis like this?

Reply:
I use this while [serializing](https://github.com/apache/iceberg/pull/11302/files#diff-fb7fc774bdf5fb14c7e9b4aab2ed645f5fcdafc7d98cce8619134cce53b4d931R126) `BitmapPositionDeleteIndex` in the main PR.

Reply:
Would it be better to just call this `compress` or `runLengthEncode`? I'm OK either way, i see roaring bit map calls it `runOptimize`, just feels like too broad of a name

Reply:
I do like `compress`, I'll rename.